Add an option to limit length of values of attributes and metric values (#1130)
open-telemetry/opentelemetry-specification: specification/common/common.md

both containing an array of strings to represent a mapping
`header_keys[i] -> header_values[i]`).

### Attribute Limits

Execution of erroneous code can result in unintended attributes. If there is no 
limits placed on attributes, they can quickly exhaust available memory, resulting 
in crashes that are difficult to recover from safely.

SDKs MAY be configured to truncate attribute values. By default, attribute values 
SHOULD NOT be truncated. If an SDK provides a way to set an attribute value size

> limit and the limit is set, then for each attribute value, serialized into a string,

Reviewer: I really think we should only be truncating string-valued attributes, based on character length, rather than having this "serialized into a string" clause. it doesn't make sense to truncate booleans, or numbers, as mentioned by another reviewer.

Author: I think we can revisit this. Right now the spec says that the possible types of values are: string, boolean, number, or a homogenous array of either of these three. So we agree that truncating strings is acceptable, even though they could be storing something serialized, like JSON. Obviously, bools and numbers will never be truncated. That leaves us with arrays. The goal of this change in the spec is to reduce COGS - so truncating arrays makes sense from this perspective, is there a case against doing so?

Reviewer: My opinion is still that we shouldn't introduce some kind of ad-hoc serialization of arrays in order to perform truncation of them. I would very much prefer that we don't try to tackle limiting array-valued attributes in this PR. If we want to try to figure out some sort of scheme for doing that, I think it would be better to have it be a separate PR/issue, as I think handling of arrays (or binary data) is much more complex and deserves its own independent discussion.

Author: Okay, I changed it to only truncate strings, and in case of string arrays to only truncate the individual items which are too long.

> if it exceeds that limit, SDK Spans MUST truncate that value, so that its length

Reviewer 2: How should this truncation be performed? From the head, tail, or random elements dropped?

Reviewer: doesn't "truncate" usually imply removing the end of the string?

> if it exceeds that limit, SDK Spans SHOULD truncate that value, so that its length

Reviewer 3: ```suggestion
if it exceeds that limit, SDK Spans MUST truncate that value, so that its length
```

is at most equal to the limit. However if a serialized value does not exceed the

> size limit, then it shouldn't be serialized.

Reviewer 3: ```suggestion
size limit, then it SHOULD NOT be serialized.
```
?

> SDKs SHOULD choose any serialization protocol, which is performant and appropriate

Reviewer 2: ```suggestion
SDKs SHOULD choose any serialization protocol which is performant and appropriate
```

for the language and/or environment they are implemented in. Please note that the

> aforementioned behavior MAY yield a previously unexpected type. For example, if 

Reviewer: I think this is a very bad idea. We should never change the type of someone's attribute.

an attribute was set and its value was an array, then after serialization, due to

> an exceeded size limit, its value SHOULD be a string. 

Reviewer 3: Where this came from? :) In Java SDK we implement attribute value truncation by truncating each individual value in an array, if it exceeds the limits. We don't try to cut the array itself.

Reviewer 4: If we say that the attribute value size limit is x, and an array is the value, then I'd presume that the size of the value is the size of the whole array. I'd be surprised if it applied to items within the array. But sure, we can reword it so that it's clear that for arrays it's applied per cell.

Author: If we say that the attribute value size limit is x, and an array is the value, then I'd presume that the size of the value is the size of the whole array. I'd be surprised if it applied to items within the array. But sure, we can reword it so that it's clear that for arrays it's applied per cell.

Reviewer 3: > we can reword it so that it's clear that for arrays it's applied per cell

I think this is a right thing to do.

Reviewer 5: > In Java SDK we implement attribute value truncation by truncating each individual value

I think this is safer approach. If we truncate array in the middle it won't be possible to parse this back from string to json / array and the whole attribute value is gone.

Author: Sure, but we still have one issues outstanding then, which is that you might add too many items to an array accidentally,  so should we have another separate limit for that?

Reviewer 5: I think it make sense

> There SHOULD be a log emitted to indicate to the user that an attribute was truncated.

Reviewer: I don't think we should recommend logging in this particular piece of the spec. We should have general logging recommendations, especially emphasizing the "do no harm" aspect. I don't think it's something that should specifically be in here, though.

Author: If we don't log anything, then you won't learn that truncation happens until you go look into a particular piece of data and you find out it is missing. What if we added a piece of metadata stating that at least one attribute value of a span was truncated, and then backend can present it as it wishes - this could be as simple as saying that if the last character of a value is an ellipsis (U+2026) then it was truncated.

Reviewer: If the user has configured their SDK to do truncation, why would they be surprised if values were being truncated?

Author: Alright, I checked it again, and the reason I added the part about logging is that it was already in place for span collection limits. I think that less changes = better. But we could raise a new PR to change this part about logging.

To prevent excessive logging, the log MUST NOT be emitted more than once per item

> on which an attribute is set.

Reviewer 2: This seems idiomatic of a language that has first class logging. I don't think it applies well universally and should not be a recommendation, but perhaps an option.

If we are specifying logging should be not be at a one-to-one cadence, can we also provide a positive recommendation? Like, once per ten items, once per minute?

Author: Again, since you've added this comment, @jkwatson said that, we should not require/recommend any logging here. Instead, it should be covered by general logging guidelines.

To define a limit SDKs SHOULD honor the environment variables specified in 
[SDK environment variables](../sdk-environment-variables.md#attribute-limits). If an
SDK does not implement truncation mechanism for all implementations of attributes,
then it SHOULD NOT offer a single global environment variable for such a limit.

> SDKs SHOULD only offer environment variables for these types of attributes, for
> which that SDK implements truncation mechanism.

Reviewer 2: ```suggestion
SDKs SHOULD only offer environment variables for the types of attributes
that SDK implements truncation mechanism.
```

Due to a possible increased implementation complexity, attribute value size limit

> MUST NOT be set to any number lower than 32. Please note that certain very low limits

Reviewer 2: Attribute sizes need to be specified with a unit here. Does 32 represent bytes, bits, characters, runes ...

Also, I'm guessing this is implying that this is a configured value (not the default), what should an SDK do if they are configured with a size smaller than this? Ignore it and revert to a default, or maybe just don't truncate?

Author: Since your comment there's been many voices to not truncate non-string values, so this particular part doesn't make sense anymore.

could even result in truncation of small numbers or booleans, and such a behavior

> is not valuable for preventing SDKs from exhausting large amounts of memory.

Reviewer 3: Although I don't oppose this, I wonder when is this problem possible? I mean in Java we apply limits just to String-typed attributes, not to others. Is this a problem for some specific language?

Author: It came up during the review of implementation in JS, so as I understand it specs should clearly state what the expectations are, but also common sense. If someone sets a limit of 4 characters, then it should truncate `false` to "fals" and `1e100/3` to '3.333', which is ridiculous.

Reviewer 5: I would argue if such limit is needed. We should not think if it makes sense to have `fals` instead of `false` as we are creating sdk which gives user full control over it and user should be able to decide if that makes sense for him or not. 

Some analogy: if in linux you want to delete all folders as a root you can do it, it won't ask you or block you from doing it, in windows it won't be possible. I prefer linux approach because it is a final user who decides about it according to needs.

I'm wondering if we should think of splitting this into 3 categories and allow user to decide if length of attribute should be truncated based on its type. This way user would have full control and could decide not to truncate boolean, not to truncate numbers and only truncate strings. So in fact this limit would only apply to string values by default but user could still override it for numbers and boolean - if for any reason he wants to do it like this (doesn't matter if that makes sense or not, for user it can make sense)

Author: That sounds reasonable, let me amend the PR.

both containing an array of strings to represent a mapping
`header_keys[i] -> header_values[i]`).

### Attribute Limits

Execution of erroneous code can result in unintended attributes. If there is no
limits placed on attributes, they can quickly exhaust available memory, resulting
in crashes that are difficult to recover from safely.

By default an SDK SHOULD apply truncation as per the list of
[configurable parameters](#attribute-limits-configuration) below.

If an SDK provides a way to:

- set an attribute value length limit and the limit is set, then for each
  attribute value:
  - if it is a string, if it exceeds that limit (counting any character in it as
    1), SDK Spans MUST truncate that value, so that its length is at most equal
    to the limit,
  - if it is an array of strings, then apply the above rule to each of the
    values separately,
  - otherwise a value MUST NOT be truncated;
- set a limit of unique attribute keys, and the limit is set:
  - for each unique attributes key, addition of which would result in exceeding
    the limit, SDK MUST discard that key/value pair.

> There SHOULD be a log emitted to indicate to the user that an attribute was

Reviewer 6: I agree with the sentiment of this paragraph -- we should tell the user when we discard or modify their telemetry data and we should be careful not to overwhelm a telemetry collection system by providing too many error messages about bad data.

This, to me, is the wrong place to specify error handling behavior. I'd like to see statements about error handling in general in the [library guidelines](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/library-guidelines.md) part of the specification. For errors like this which may occur as a result of user data, I typically configure them to print once per minute or so. I believe we should use metrics to count occurrences of excessive metric value length and/or count. 

Both of these outputs, logs and/or metrics about these events, are a form of "meta" telemetry, and the specification's SDK guidelines should say we aim to (1) do this well, (2) at low cost. For meta-telemetry logs this means once rarely per log statement, for metrics this means a requirement to have low cardinality, for example.

I propose removing this paragraph and filing an issue to update the library guidelines with general guidance on handling errors of this nature.

Author: Given that this is in line with existing spec here <https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#span-limits>, do you think we could leave the paragraph as is, create the issue, and once we reach a consensus, I'll create a PR which will address this?

Author: @jmacd please feel free to amend the issue #1787 if you'd like to describe the problem in a different way

Author: @jmacd does the above work for you?

Reviewer 6: Yes.

Reviewer 7: I think solving this issue with logging is not the ideal solution. It assumes that logging is always available, which is not always the case - many languages have no standard log framework, and even if they do, using std log is not always what the rest of the application is using. In other words, logger must always be treated as a dependency, which may or may not be available.

In addition, the throttling of the logs effectively requires keeping a state per span. So why not report truncations as part of the span itself?

Regarding #1787, why not leave this paragraph out of this PR until an agreement on #1787 is reached? We are changing the spec, after all, it's better to leave the behavior unspecified than to change it.

Author: @yurishkuro I have changed "SHOULD" to "SHOULD NOT" to satisfy this criteria:
- your ask to not prescribe a solution
- don't immediately invalidate existing solutions, which may already be logging
- don't make it hard for implementors in case they really needed to log (until we have a prescribed solution)

This isn't exactly what you asked for, but I think it does the job

both containing an array of strings to represent a mapping
`header_keys[i] -> header_values[i]`).

### Attribute Limits

Execution of erroneous code can result in unintended attributes. If there is no
limits placed on attributes, they can quickly exhaust available memory, resulting
in crashes that are difficult to recover from safely.

SDKs MAY be configured to truncate attribute lists and/or their values. By
default, attribute lists and values SHOULD NOT be truncated.

If an SDK provides a way to:

> - set an attribute value length limit and the limit is set, then for each

Reviewer 2: What unit is a limit define in? Bytes?

Author: I think the statement a couple lines above
> (counting any character in it as 1)

is understandable.

> - set an attribute value length limit and the limit is set, then for each

Reviewer 2: Is there planned a limit on the size an attribute key can be?

Author: That doesn't seem like a risk worth addressing. For this to happen you'd need to have a complex logic of generating keys. That said, maybe someone with longer instrumentation tenure can chime in.

  attribute value:
  - if it is a string, if it exceeds that limit (counting any character in it as

>     1), SDK Spans MUST truncate that value, so that its length is at most equal

Reviewer 2: Counting a character as 1 is not in line with the purpose of adding limits to protect memory usage. This should likely be defined in terms of memory used as character width is not universal.

Reviewer: I don't disagree, but if we had to implement byte-counting for characters in java, that would be tricky, as it would depend on the encoding to use, which isn't really known until export time.

Reviewer 2: Hmm. @jkwatson does character counting the best we can do then in Java (and I'm guessing in other languages)?

Is it possible to use the maximum character width allowed (if that is possible to know?) multiplied by the characters?

Reviewer: maximum character width is generally 4 bytes (I think?), so that would be 4x what people are normally going to use for this stuff, which is probably standard ASCII. I'd suggest just counting characters and users who know that they are using multi-byte characters frequently could adjust their limits accordingly.

Reviewer 2: > maximum character width is generally 4 bytes (I think?), so that would be 4x what people are normally going to use for this stuff, which is probably standard ASCII. I'd suggest just counting characters and users who know that they are using multi-byte characters frequently could adjust their limits accordingly.

:+1: sounds like a plan

    to the limit,
  - if it is an array of strings, then apply the above rule to each of the
    values separately,

>   - otherwise a value MUST NOT be truncated;

Reviewer 2: If a limit is set smaller than a numeric value can be represented in (`float64` or `int64`), what is to be done there?

Author: There was a discussion about this earlier in this PR, and that's why non-strings are explicitly excluded.

> - set a limit of unique attribute keys, and the limit is set:

Reviewer 8: This is a complex topic.
Do we set the limit on the SDK instance level, or provider level (e.g. if there are two TracerProviders and one provider is going crazy, do we want the other one to work independently? - my personal answer is YES)

Author: It's stated elsewhere that the limit is per data record on which the attributes are set, which to me implies that they are independent.

  - for each unique attributes key which exceeds a limit, SDK MUST discard that

>     key/value pair.

Reviewer 2: Is this defined as FIFO or FILO? The way I read this currently is sounds like the whole set is dropped.

Author: I don't know if queue terminilogy applies that well here, but hopefully it's clear now:
> [...] addition of which [attribute] would result in exceeding the limit, SDK MUST discard that key/value pair

There SHOULD be a log emitted to indicate to the user that an attribute was

> truncated. To prevent excessive logging, the log MUST NOT be emitted more than

Reviewer 2: Should a log message be emitted when an attribute key/value pair is dropped as well?

Author: Yes, done.

once per record on which an attribute is set. If a record is embedded in another
record (e.g. Events within a Span), then the log SHOULD NOT be emitted more than

> once per parent record.

Reviewer 2: What is the top level "parent record" here? All spans are in a trace, is that a valid parent?

If the SDK implements the limits above, it MUST provide a way to change these

> limits programatically. Names of the configuration options SHOULD be the same as

Reviewer 8: ```suggestion
limits programmatically. Names of the configuration options SHOULD be the same as
```

in the list below. Implementations MAY provide additional configuration options

> such as `AttributePerEventCountLimit` or `AttributePerLinkCountLimit`.

Reviewer 2: If we plan to allow these, they need to be specified. Otherwise, there is potential for different implementations to use different values.

> <a name="attribute-limits-configuration"></a>

Reviewer 2: Why the HTML and not a markdown link?

Author: Can I create an anchor in pure markdown? I thought [this was how everyone does it](https://stackoverflow.com/questions/5319754/cross-reference-named-anchor-in-markdown).

**Configurable parameters:**

> * `AttributeCountLimit` (Default=128) - Maximum allowed attribute count per record;

Reviewer 2: Having a default of 128 is at odds with the statement:

> By default, attribute lists and values SHOULD NOT be truncated.

Author: Fixed.

Reviewer: I don't think we should have a default value for this. I think we should stay with the current behavior as the default, which is no truncation. Otherwise people are going to have a surprising potentially breaking change to their telemetry when they upgrade to a version of the SDK that has a default.

Author: @jkwatson This default is the same as the one which currently exists for spans, and metrics are exempt per the ask from Reiley.

Reviewer: 🤦🏼  sorry, you are correct!

in crashes that are difficult to recover from safely.

SDKs MAY be configured to truncate attribute lists and/or their values. By
default, attribute lists and values SHOULD NOT be truncated.

If an SDK provides a way to:

- set an attribute value length limit and the limit is set, then for each
  attribute value:
  - if it is a string, if it exceeds that limit (counting any character in it as
    1), SDK Spans MUST truncate that value, so that its length is at most equal
    to the limit,
  - if it is an array of strings, then apply the above rule to each of the
    values separately,
  - otherwise a value MUST NOT be truncated;
- set a limit of unique attribute keys, and the limit is set:
  - for each unique attributes key which exceeds a limit, SDK MUST discard that
    key/value pair.

There SHOULD be a log emitted to indicate to the user that an attribute was
truncated. To prevent excessive logging, the log MUST NOT be emitted more than
once per record on which an attribute is set. If a record is embedded in another
record (e.g. Events within a Span), then the log SHOULD NOT be emitted more than
once per parent record.

> To define a limit SDKs SHOULD honor the environment variables specified in

Reviewer: I don't think we should mix up the *what* with the *how*. The primary thing is to make sure the SDK has support for the feature. That support *must* be programmatic at the very least. We should describe the *how* of environment variable support in the documentation of environment variables, not here.

Author: Relevant parts moved to ENV variables doc.

[SDK environment variables](../sdk-environment-variables.md#attribute-limits).
If an SDK does not implement truncation mechanism for all implementations of
attributes, then it SHOULD NOT offer an unscoped (e.g.
`OTEL_ATTRIBUTE_VALUE_LENGTH_LIMIT`, as opposed to e.g.
`OTEL_SPAN_ATTRIBUTE_VALUE_LENGTH_LIMIT`) global environment variable for such a
limit. SDKs SHOULD only offer environment variables for the types of attributes,
for which that SDK implements truncation mechanism. For each implemented limit
SDK MUST first try to use the unscoped variable, if it isn't set, only then try
to use the scoped variable.

If the SDK implements the limits above, it MUST provide a way to change these
limits via a configuration passed to the to the `<Type>Provider`, where `<Type>`
is any kind of a producer of records, which implement attributes.

The options MAY be bundled in a class, which then SHOULD be called

> `<Record>Limits` (where `<Record>` is a type of a record, e.g. `Span`). Names of

Reviewer: I don't think we should bother to mention this; it's an implementation detail that is language-specific and generally not relevant for the spec.

Author: done

the configuration options SHOULD be the same as in the list below.
Implementations MAY provide additional configuration options such as
`AttributePerEventCountLimit` or `AttributePerLinkCountLimit`.

<a name="attribute-limits-configuration"></a>
**Configurable parameters:**

* `AttributeCountLimit` (Default=128) - Maximum allowed attribute count per record;
* `AttributeValueLengthLimit` (Default=Infinity) - Maximum allowed attribute value length;

SDKs MAY use a different notation (e.g. snake_case) if it's more suitable for

> its language.

Reviewer: This is also not relevant, as it is assumed to be the case for any SDK options. I'd remove it.

Author: done

`header_keys[i] -> header_values[i]`).

See [Attribute Naming](attribute-naming.md) for naming guidelines.

### Attribute Limits

Execution of erroneous code can result in unintended attributes. If there is no
limits placed on attributes, they can quickly exhaust available memory, resulting
in crashes that are difficult to recover from safely.

By default an SDK SHOULD apply truncation as per the list of
[configurable parameters](#attribute-limits-configuration) below.

If an SDK provides a way to:

- set an attribute value length limit and the limit is set, then for each
  attribute value:
  - if it is a string, if it exceeds that limit (counting any character in it as

>     1), SDK Spans MUST truncate that value, so that its length is at most equal

Reviewer 2: nit: it is the SDK that does the truncation.

```suggestion
    1), SDKs MUST truncate that value, so that its length is at most equal
```

Author: If I remember correctly, this is to specify that the truncation is done at the Span level and not elsewhere, but @carlosalberto feel free to make this call when merging.

`header_keys[i] -> header_values[i]`).

See [Attribute Naming](attribute-naming.md) for naming guidelines.

### Attribute Limits

> Execution of erroneous code can result in unintended attributes. If there is no

Reviewer: s/is/are/

Author: thanks!

limits placed on attributes, they can quickly exhaust available memory, resulting
in crashes that are difficult to recover from safely.

By default an SDK SHOULD apply truncation as per the list of
[configurable parameters](#attribute-limits-configuration) below.

If an SDK provides a way to:

> - set an attribute value length limit and the limit is set, then for each

Reviewer 7: ```suggestion
- set an attribute value length limit such that for each
```

  attribute value:
  - if it is a string, if it exceeds that limit (counting any character in it as
    1), SDKs MUST truncate that value, so that its length is at most equal
    to the limit,
  - if it is an array of strings, then apply the above rule to each of the
    values separately,
  - otherwise a value MUST NOT be truncated;

> - set a limit of unique attribute keys, and the limit is set:

Reviewer 7: ```suggestion
- set a limit of unique attribute keys such that:
```

  - for each unique attributes key, addition of which would result in exceeding
    the limit, SDK MUST discard that key/value pair.

> There SHOULD NOT be a log emitted to indicate to the user that an attribute was

Reviewer 9: ```suggestion
There SHOULD be a log emitted to indicate to the user that an attribute was
```

Author: This is not a mistake, see https://github.com/open-telemetry/opentelemetry-specification/pull/1130#discussion_r671702102

Reviewer 9: Aah! I see it now :)
When read in combination with the next sentence, it feels like "you should log, but be careful to not do excessive log."

I think your intent is "You should not log. (not a Must, so libraries may still do the logging). But if you decide to log anyway, you *must* not do excessive log"

I'll resolve this comment!

> There SHOULD NOT be a log emitted to indicate to the user that an attribute was

Reviewer 7: ```suggestion
There MAY be a log emitted to indicate to the user that an attribute was
```

Author: Okay, if you prefer it this way.

- set a limit of unique attribute keys such that:
  - for each unique attributes key, addition of which would result in exceeding
    the limit, SDK MUST discard that key/value pair.

There MAY be a log emitted to indicate to the user that an attribute was
truncated or discarded. To prevent excessive logging, the log MUST NOT be
emitted more than once per record on which an attribute is set.

If the SDK implements the limits above, it MUST provide a way to change these
limits programmatically. Names of the configuration options SHOULD be the same as
in the list below.

An SDK MAY implement model-specific limits, for example
`SpanAttributeCountLimit`. If both a general and a model-specific limit are
implemented, then the SDK MUST first attempt to use the model-specific limit, if
it isn't set and doesn't have a default, then the SDK MUST attempt to use the
general limit.

<a name="attribute-limits-configuration"></a>
**Configurable parameters:**

* `AttributeCountLimit` (Default=128) - Maximum allowed attribute count per record;
* `AttributeValueLengthLimit` (Default=Infinity) - Maximum allowed attribute value length;

#### Exempt Entities

> Attributes which belong to Metrics MUST NOT implement the aforementioned

Reviewer 7: again, you cannot put MUST's into spec if the intention is to change the rules in the near future. It's not a backwards compatible change. Either leave the behavior unspecified or use the weak MAY.
